text_generation/causal_lm/cpp/speculative_decoding_lm.cpp

```diff
@@ -0,0 +1,294 @@
+    OPENVINO_ASSERT(seq_len_axis == 2 || seq_len_axis == 0, "Cannot trim key/values with sequence length axis = ", seq_len_axis);
+    
+    auto old_tensor_data = tensor.data<float>();
+    auto shape = tensor.get_shape();
+    size_t size_1 = shape[1];
+    size_t old_seq_len = shape[2];
+    size_t size_2 = shape[3];
+    
+    // if new_seq_len equal to old one no need to copy tensor, return as is
+    if (old_seq_len == new_seq_len)
+        return tensor;
+    
+    // if seq_len_axis is the very first dimension, this means data is contiguous, then trim by just setting the new shape
+    if (seq_len_axis == 0) {
+        shape[0] = new_seq_len;
+        tensor.set_shape(shape);
+        return tensor;
+    }
+
+    // if seq_len_axis == 2, then data is not contiguous, in order to trim need to repack tensor
+    auto new_tensor = ov::Tensor{ov::element::f32, {BATCH_SIZE, size_1, new_seq_len, size_2}};
+    auto new_tensor_data = new_tensor.data<float>();
+    for (size_t batch = 0; batch < BATCH_SIZE; ++batch){
+        for (size_t i = 0; i < size_1; ++i) {
+            for (size_t j = 0; j < new_seq_len; ++j) {
+                auto dst_ptr = new_tensor_data + size_1 *new_seq_len * size_2 * batch + new_seq_len * size_2 * i +  size_2 * j;
+                auto src_ptr = old_tensor_data + size_1 *new_seq_len * size_2 * batch + old_seq_len * size_2 * i +  size_2 * j;
+                std::memcpy(dst_ptr, src_ptr, size_2 * sizeof(float));
+            }
+        }
+    }
+    return new_tensor;
+}
+
+void update_kv_cache(ov::InferRequest request, uint64_t seq_len_axis, uint64_t new_seq_len) {
+    // trims kv_cache values up to the new_seq_len
+    auto num_outputs = request.get_compiled_model().outputs().size();
+    for (size_t idx = 1; idx < num_outputs; ++idx) {
+        auto old_tensor = request.get_output_tensor(idx);
+        auto trimmed_tensor = trimm_tensor(old_tensor, seq_len_axis, new_seq_len);
+        request.set_output_tensor(idx, trimmed_tensor);
+    }
+}
+
+int main(int argc, char* argv[]) try {
+    if (argc != 4) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <DRAFT MODEL_DIR> <MAIN MODEL_DIR> '<PROMPT>'");
+    }
+
+    // tokenizer model
+    ov::Core core;
+    core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
+    // tokenizer and detokenizer work on CPU only
+    ov::InferRequest tokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_tokenizer.xml", "CPU").create_infer_request();
+    auto [draft_input_ids, draft_attention_mask] = tokenize(tokenizer, argv[3]);
+    ov::InferRequest detokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    TextStreamer text_streamer{std::move(detokenizer)};
+
+    // draft model
+    ov::InferRequest draft_model = core.compile_model(std::string{argv[1]} + "/openvino_model.xml", "CPU").create_infer_request();
```

Comment:
@AlexKoff88 in HG it's called assistant model (see https://huggingface.co/blog/assisted-generation#greedy-decoding-with-assisted-generation)
is `draft_model` our invention or common name?

Reply:
in the original paper it's called draft model, so it's not our invention it's a common name
https://arxiv.org/pdf/2302.01318.pdf